include/aws/iotdevice/private/secure_tunneling_operations.h

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#ifndef AWS_IOTDEVICE_SECURE_TUNNELING_OPERATION_H
+#define AWS_IOTDEVICE_SECURE_TUNNELING_OPERATION_H
+
+#include <aws/common/byte_buf.h>
+#include <aws/common/linked_list.h>
+#include <aws/common/ref_count.h>
+#include <aws/iotdevice/private/serializer.h>
+#include <aws/iotdevice/secure_tunneling.h>
+
+/*********************************************************************************************************************
+ * Operations
+ ********************************************************************************************************************/
+
+struct aws_secure_tunnel_operation;
+
+enum aws_secure_tunnel_operation_type {
+    AWS_STOT_NONE,
+    AWS_STOT_PING,
+    AWS_STOT_DATA,
+    AWS_STOT_STREAM_RESET,
+    AWS_STOT_STREAM_START
+};
+
+struct aws_secure_tunnel_message_storage {
+    struct aws_allocator *allocator;
+    struct aws_secure_tunnel_message_view storage_view;
+
+    bool ignorable;
+    int32_t stream_id;
+    struct aws_byte_cursor service_id;
+    struct aws_byte_cursor payload;
+
+    struct aws_byte_buf storage;
+};
+
+/* Basic vtable for all secure tunnel operations.  Implementations are per-message type */
+struct aws_secure_tunnel_operation_vtable {
+    void (*aws_secure_tunnel_operation_completion_fn)(
+        struct aws_secure_tunnel_operation *operation,
+        int error_code,
+        const void *completion_view);
+
+    /* Set the stream id of outgoing st_msg */
+    int (*aws_secure_tunnel_operation_set_stream_id_fn)(
+        struct aws_secure_tunnel_operation *operation,
+        struct aws_secure_tunnel *secure_tunnel);
+
+    /* Set the stream id of outgoing st_msg to +1 of the currently set stream id */
+    int (*aws_secure_tunnel_operation_set_next_stream_id_fn)(
+        struct aws_secure_tunnel_operation *operation,
+        struct aws_secure_tunnel *secure_tunnel);
+};
+
+/**
+ * This is the base structure for all secure tunnel operations.  It includes the type, a ref count, and list management.
+ */
+struct aws_secure_tunnel_operation {
+    const struct aws_secure_tunnel_operation_vtable *vtable;
+    struct aws_ref_count ref_count;
+    struct aws_linked_list_node node;
+
+    enum aws_secure_tunnel_operation_type operation_type;
+    const struct aws_secure_tunnel_message_view *message_view;
+
+    /* Size of the secure tunnel message this operation represents */
+    size_t message_size;
+
+    void *impl;
+};
+
+struct aws_secure_tunnel_operation_message {
+    struct aws_secure_tunnel_operation base;
+    struct aws_allocator *allocator;
+
+    struct aws_secure_tunnel_message_storage options_storage;
+};
+
+struct aws_secure_tunnel_operation_pingreq {
+    struct aws_secure_tunnel_operation base;
+    struct aws_allocator *allocator;
+};
+
+/*
+ * Secure tunnel configuration
+ */
+struct aws_secure_tunnel_options_storage {
```

Comment:
I'll take a closer look at whether we should move this to impl. At one point I was thinking about removing secure_tunneling_impl to reduce the numebr of private header files and this might be a remnant of that.